### PR TITLE
Persistence Layer For RestEndpoint phase one

### PR DESCRIPTION
**Pull request by SarpongAbasimi**
# What I have done 

- Created a base trait for the transactor as suggested
- Created an Algebra and Imp for SubscriptionService
- Altered the db schema to add a subscribed at column 

This Pr is to get feedback on what has currently been done so that it can be improved before progressing further.

# Also I came across this issue today. 
- It is possible to do `query.[GetSubscriptionData]` (which is what I currently have)
- But  I realised that I actually needed a `List[GetSubscriptionData]` which is what `GetSubscriptions` is.
- Getting this error when trying to build the application ( I am looking into)

# After 300 years of debugging things finally work as expected 😂

#### What I did

- Altered `db` to generate `uuid` by default
- Added test to test the endpoint (Can't currently test the output json as the subscribe_at time is generated automatically)
- Removed mapping of `subscribe_at` to `varchar`

I would appreciate a final review from @fedefernandez or @moleike before merging this.

# Next

- I will probably add some more test and think about edge cases
- Then progress with the next ticket.

### CODE REVIEW

File: modules/algebrasAndModel/src/main/scala/subscriptionAlgebra/SubscriptionServiceAlgebra.scala [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package subscriptionAlgebra
+
+import utils.Types.{GetSubscriptionData, Id, PostSubscriptions}
+
+trait SubscriptionServiceAlgebra[F[_]] {
+  def get(id: Id): F[Option[GetSubscriptionData]]
```
Comment (SarpongAbasimi) [inline modules/algebrasAndModel/src/main/scala/subscriptionAlgebra/SubscriptionServiceAlgebra.scala]: I just figured out that this should probably be the `slackId` because the ID being passed to the the Url is a String.

File: modules/persistenceService/src/main/resources/db/migration/V3__ADD_SUBS_AT_FIELD.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE subscriptions
+ADD COLUMN subscribed_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP;
```
Comment (moleike) [inline modules/persistenceService/src/main/resources/db/migration/V3__ADD_SUBS_AT_FIELD.sql]:
```suggestion
ADD COLUMN subscribed_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP;
```
Or work with `Option` if you want to make the column nullable.

File: modules/algebrasAndModel/src/main/scala/subscriptionAlgebra/SubscriptionServiceAlgebra.scala (hunk 1 above)
Comment (moleike) [reply modules/algebrasAndModel/src/main/scala/subscriptionAlgebra/SubscriptionServiceAlgebra.scala]: Yes, you'd probably want to have a relation with the one-to-one mapping between your internal ids and the upstream services like Slack. Let me know if you need help with this.

File: modules/algebrasAndModel/src/main/scala/utils/Types.scala [hunk 3]
```diff
@@ -25,11 +25,19 @@ object Types {
...
   )
-  final case class Organization(organization: String) extends AnyVal
-  object Organization {
-    implicit val organizationEncoder: Encoder[Organization] = deriveUnwrappedEncoder[Organization]
-    implicit val organizationDecoder: Decoder[Organization] = deriveUnwrappedDecoder[Organization]
+
+  final case class Owner(owner: String) extends AnyVal
+  object Owner {
+    implicit val encoder: Encoder[Owner] = deriveUnwrappedEncoder[Owner]
+    implicit val decoder: Decoder[Owner] = deriveUnwrappedDecoder[Owner]
```
Comment (moleike) [inline modules/algebrasAndModel/src/main/scala/utils/Types.scala:33]: Nitpick: you could also either use circe Codec (`trait Codec[A] extends Decoder[A] with Encoder[A]`) to make instance definition more concise.

*Below is not necessary for you to do, it's just FYR*

If you want to go the extra mile, you could use @newtype macro instead of value classes and  derevo @derive macro annotation:

```scala
@derive(encoder, decoder)
@newtype
case class Owner(owner: String)
```

File: modules/persistenceService/src/main/resources/db/migration/V2__ALTER_USER_TABLE_DROP_ADD_NEW_COLUMN.sql [hunk 4]
```diff
@@ -12,9 +12,9 @@ CREATE TABLE repositories(
 );
 
 CREATE TABLE subscriptions(
-    id uuid NOT NULL,
+    user_id uuid NOT NULL,
```
Comment (moleike) [inline modules/persistenceService/src/main/resources/db/migration/V2__ALTER_USER_TABLE_DROP_ADD_NEW_COLUMN.sql]: here I think it'd be better to add a versioned migration:

```sql
alter table subscriptions rename column id to user_id;
```

File: modules/subscriptionService/src/main/scala/service/SubscriptionService.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+import cats.effect.Sync
+import userDbAlgebra.UserAlgebra
+import subscriptionAlgebra.SubscriptionServiceAlgebra
+import utils.Types.{GetSubscriptionData, Id, PostSubscriptions}
+import doobie.ConnectionIO
+import doobie.implicits._
+import doobie.util.transactor.Transactor
+
+sealed trait UserErrors                  extends Exception
```
Comment (moleike) [inline modules/subscriptionService/src/main/scala/service/SubscriptionService.scala]: Suggestion: move this to a different package with domain alebraic data types.

File: modules/subscriptionService/src/main/scala/service/SubscriptionService.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+              UserNotFound("User does not Exit")
+            )
+          case Some(user) => subscriptionAlgebra.get(user.id)
+        }
+      } yield optionOfGetSubscription: Option[GetSubscriptionData]
+    }.transact(transactor)
+
+    def post(id: Id, subscriptions: PostSubscriptions): F[Unit] = (for {
+      checkIfTheUserExits <- userAlgebra.findUser(id)
+      _ <- checkIfTheUserExits match {
```
Comment (moleike) [inline modules/subscriptionService/src/main/scala/service/SubscriptionService.scala]: Did you solve the problem with the inferred `Any`?

```scala
  "-Xlint:infer-any",                  // Warn when a type argument is inferred to be `Any`.
```
I am not sure if you are using it already, but I highly recommend using this plugin: 

```scala
addSbtPlugin("io.github.davidgregory084" % "sbt-tpolecat"        % "0.1.20")
```

File: modules/persistenceService/src/test/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      .unsafeRunSync()
+  }
+
+  describe("SubscriptionServiceQuery") {
+    describe("get") {
+      describe("when called") {
+        it("should return a list of repositories to the user") {
+          val id    = Id(UUID.randomUUID())
+          val query = SubscriptionServiceQuery.get(id)
+          check(query)
```
Comment (moleike) [inline modules/persistenceService/src/test/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayerSpec.scala:33]: I see you are type checking the queries, good. But is there somewhere in the test where you actually exercise the database? If not yet, it'd be good to add it in a follow-up PR.

**Review by moleike — APPROVED** [review]
Great stuff Sarpong. I sprinkled some comments.

File: modules/algebrasAndModel/src/main/scala/subscriptionAlgebra/SubscriptionServiceAlgebra.scala (hunk 1 above)
Comment (SarpongAbasimi) [reply modules/algebrasAndModel/src/main/scala/subscriptionAlgebra/SubscriptionServiceAlgebra.scala]: Yeah I will definitely reach out yo You on this one. Thanks @moleike

File: modules/persistenceService/src/main/resources/db/migration/V2__ALTER_USER_TABLE_DROP_ADD_NEW_COLUMN.sql (hunk 4 above)
Comment (SarpongAbasimi) [reply modules/persistenceService/src/main/resources/db/migration/V2__ALTER_USER_TABLE_DROP_ADD_NEW_COLUMN.sql]: sure will do

File: modules/persistenceService/src/test/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayerSpec.scala (hunk 7 above)
Comment (SarpongAbasimi) [reply modules/persistenceService/src/test/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayerSpec.scala:33]: Yeah this Pr was actually to get Feedback before the Pr before too big.
I will make all the changes now. Thank you

File: modules/subscriptionService/src/main/scala/service/SubscriptionService.scala (hunk 6 above)
Comment (SarpongAbasimi) [reply modules/subscriptionService/src/main/scala/service/SubscriptionService.scala]: @moleike  I resolved it by specifying the return type on the `yield`.
Also I will have a loot at the plugin thank you 😊

File: modules/persistenceService/src/main/resources/db/migration/V2__ALTER_USER_TABLE_DROP_ADD_NEW_COLUMN.sql [hunk 8]
```diff
@@ -15,6 +15,6 @@ CREATE TABLE subscriptions(
     id uuid NOT NULL,
     repository_id uuid NOT NULL,
     PRIMARY KEY(id, repository_id),
-    FOREIGN KEY (id) REFERENCES userdb(id) ON UPDATE CASCADE,
-    FOREIGN KEY (repository_id) REFERENCES repositories(repository_id) ON UPDATE CASCADE
+    FOREIGN KEY(id) REFERENCES userdb(id) ON UPDATE CASCADE,
+    FOREIGN KEY(repository_id) REFERENCES repositories(repository_id) ON UPDATE CASCADE
```
Comment (fedefernandez) [inline modules/persistenceService/src/main/resources/db/migration/V2__ALTER_USER_TABLE_DROP_ADD_NEW_COLUMN.sql]: Comment: Flyway generates a checksum of the applied migrations and then checks the value when applying the migrations. This means that we shouldn't modify any existing migration that has been merged.

File: modules/persistenceService/src/main/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayer.scala [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        } yield ()
+
+      def delete(id: Id, subscriptions: PostSubscriptions): ConnectionIO[Unit] =
+        SubscriptionServiceQuery.delete(id).updateMany(subscriptions.subscriptions).void
+    }
+}
+
+object SubscriptionServiceQuery {
+  def get(id: Id): Query0[GetSubscriptionData] =
+    sql"select owner, repository, subscribed_at::varchar from repositories join subscriptions using (repository_id) where user_id = ${id.id}"
```
Comment (fedefernandez) [inline modules/persistenceService/src/main/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayer.scala]: To be honest I'm not a big fan of the type mapping here of `subscribed_at`. What kind of error are you getting?

File: modules/persistenceService/src/main/resources/db/migration/V2__ALTER_USER_TABLE_DROP_ADD_NEW_COLUMN.sql (hunk 8 above)
Comment (SarpongAbasimi) [reply modules/persistenceService/src/main/resources/db/migration/V2__ALTER_USER_TABLE_DROP_ADD_NEW_COLUMN.sql]: ohh so even a space in the file will cause an issue. I didn't know that before. 
Thank you Fede

File: modules/persistenceService/src/main/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayer.scala (hunk 9 above)
Comment (SarpongAbasimi) [reply modules/persistenceService/src/main/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayer.scala]: Thanks @fedefernandez 
Please this is because my `case class` expects `subscribed_at` to be of type `String`.
`ADD COLUMN subscribed_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP;`

```
[info]           TIMESTAMP (timestamp) is ostensibly coercible to String according
[info]           to the JDBC specification but is not a recommended target type.
[info]           Expected schema type was CHAR or VARCHAR or LONGVARCHAR or NCHAR
[info]           or NVARCHAR or LONGNVARCHAR. (Checker.scala:56)
```

Screenshot 2021-08-17 at 18 45 54

Comment (moleike) [reply modules/persistenceService/src/main/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayer.scala]: How about this:

```diff
+ import java.time.OffsetDateTime

- final case class SubscribeAt(subscribeAt: String) extends AnyVal
+ final case class SubscribeAt(value: OffsetDateTime) extends AnyVal

  object SubscribeAt {
    implicit val decoder: Decoder[SubscribeAt] = deriveUnwrappedDecoder[SubscribeAt]
    implicit val encoder: Encoder[SubscribeAt] = deriveUnwrappedEncoder[SubscribeAt]
  }
```

This way you don't need a typecast, plus you work with time objects in your code, safer and more versatile than raw strings.

Comment (SarpongAbasimi) [reply modules/persistenceService/src/main/scala/subsPersistenceLayer/SubscriptionServicePersistenceLayer.scala]: Thank you very much @moleike

File: modules/subscriptionService/src/main/scala/service/SubscriptionService.scala (hunk 6 above)
Comment (SarpongAbasimi) [reply modules/subscriptionService/src/main/scala/service/SubscriptionService.scala]: @moleike I am not having a lot of success with this plugin 😂

**Review by moleike — APPROVED** [review]
